tests/unit/amo/components/TestHeader.js

```diff
@@ -40,4 +63,95 @@ describe(__filename, () => {
     expect(root.find('.Header-title').prop('children'))
       .toContain('Firefox Add-ons');
   });
+
+  it('displays `login` text when user is not connected', () => {
+    const { store } = dispatchClientMetadata();
+    const fakeI18n = getFakeI18nInst();
+    const allProps = { i18n: fakeI18n, location: {}, query: '' };
+
+    const wrapper = mount(
+      <Provider store={store}>
+        <I18nProvider i18n={fakeI18n}>
+          <Header {...allProps} />
+        </I18nProvider>
+      </Provider>
+    );
+
+    expect(wrapper.find(AuthenticateButton).text())
+      .toEqual('Register or Log in');
+    expect(wrapper.find(DropdownMenu)).toHaveLength(0);
+  });
+
+  it('displays a menu and the username when user is logged in', () => {
+    const { store } = dispatchSignInActions({ username: 'babar' });
+    const fakeI18n = getFakeI18nInst();
+    const allProps = { i18n: fakeI18n, location: {}, query: '' };
+
+    const wrapper = mount(
+      <Provider store={store}>
+        <I18nProvider i18n={fakeI18n}>
+          <Header {...allProps} />
+        </I18nProvider>
+      </Provider>
+    );
+
+    expect(wrapper.find(DropdownMenu)).toHaveLength(1);
+    expect(wrapper.find(DropdownMenu)).toIncludeText('babar');
+  });
+
+  it('allows to log out a logged user', () => {
+    const { store } = dispatchSignInActions({ username: 'babar' });
+    const fakeI18n = getFakeI18nInst();
+    const allProps = { i18n: fakeI18n, location: {}, query: '' };
+    const mockApi = sinon.mock(api);
+
+    mockApi
+      .expects('logOutFromServer')
+      .once()
+      .returns(Promise.resolve());
+
+    const wrapper = mount(
+      <Provider store={store}>
+        <I18nProvider i18n={fakeI18n}>
+          <Header {...allProps} />
+        </I18nProvider>
+      </Provider>
+    );
+
+    wrapper.find(DropdownMenuItem).find('button').simulate('click');
+
+    mockApi.verify();
+  });
+
+  describe('mapStateToProps', () => {
+    it('gets props from store', () => {
+      const { store } = dispatchClientMetadata();
+      store.dispatch(setViewContext(VIEW_CONTEXT_EXPLORE));
+      expect(mapStateToProps(store.getState())).toEqual({
+        api: store.getState().api,
+        isAuthenticated: false,
```

Comment:
I agree with what @tofumatt said. You should not need to test `mapStateToProps` directly (in most cases) when you're using `shallowUntilTarget`. Test the mapped properties instead which will also help us cover our Redux integration. Here is an example of how to test the properties themselves:

```js
it('sets isAuthenticated from state', () => {
  const { store } = dispatchClientMetadata();
  store.dispatch(setViewContext(VIEW_CONTEXT_EXPLORE));
  const root = render({ store });
  
  expect(root.instance().props.isAuthenticated).toEqual(true);
});
```

*However*, this kind of test isn't so useful. What we really want to know is how the `isAuthenticated` prop used. It is used to show either a dropdown or an authenticate button and, at a glance, I think you already have tests to cover that logic :)

Reply:
Ok, I am fine with that. Thanks!